### ISSUE TITLE
Instrument more app start data

### ISSUE THREAD

**Issue by philipphofmann** [getsentry/sentry-cocoa#3345]
### Description

Look into Apple's [recommendations](https://developer.apple.com/documentation/xcode/reducing-your-app-s-launch-time#Move-expensive-tasks-out-of-your-app-delegate) on how to reduce the app launch time and identify if we can add more spans that measure actions our users could speed up, such as `willFinishLaunchinWithOptions` and `didFinishLaunchingWithOptions`.

**Comment by armcknight** [comment]
Check this out, wonder if we could do something similar: https://www.emergetools.com/product/launchbooster

**Comment by philipphofmann** [comment]
I played around with the [code snipped](https://gist.github.com/noahsmartin/30604f038ea805b3fe48fa0d714c959e#file-faulttiming-m) mentioned in the blog post. The code snippet measures the duration of accessing memory pages of each added binary in order or randomly. I added this code to our code base to understand how to use it and pushed the state to the branch [feat/measure-binary-page-access](https://github.com/getsentry/sentry-cocoa/compare/feat/measure-binary-page-access?expand=1). The SentryCrashDynamicLinker measures memory page access in order as the code snipped does and puts the result into SentryCrashBinaryImage. When building the app start spans, we extract this information and create spans for each binary with the duration of accessing all memory pages of the binary. The iOS-Swift sample app has around 400 binary images loaded, mostly system binaries. For a cold start, reading the pages takes time; see [transaction](https://sentry-sdks.sentry.io/discover/sentry-cocoa:029d4fec17404ac29fe177296a83279d/?field=title&field=event.type&field=project&field=user.display&field=timestamp&field=user.email&homepage=true&id=18214&name=&project=5428557&query=user.email%3Aphilipp%40example.com&sort=-timestamp&statsPeriod=1h&topEvents=5&yAxis=count%28%29). Reading the pages on a second launch is significantly faster; see [transaction](https://sentry-sdks.sentry.io/discover/sentry-cocoa:5b8a131321a845efa691d58309504a90/?field=title&field=event.type&field=project&field=user.display&field=timestamp&field=user.email&homepage=true&id=18214&name=&project=5428557&query=user.email%3Aphilipp%40example.com&sort=-timestamp&statsPeriod=1h&topEvents=5&yAxis=count%28%29).
The main problem is that reading all the memory pages of the binaries slows down the app start significantly, which is not a surprise cause a regular app launch only reads a subset of those memory pages. So this is something we could not use in production. 

Ideally, we would build something that detects and measures the order of memory access in production, but intercepting memory access seems dangerous. I would need to do further investigations if and how that is achievable. Do you have any ideas on how to do that, @armcknight?

Anyway, I don't think we can add more app start data, as pointed out in the blog post. Instead, we could add measurements for the number of external frameworks and dynamic libraries once we ad measurements to our SDK, which also directly impacts app start duration, as pointed out by the [Apple docs](https://sentry-sdks.sentry.io/discover/sentry-cocoa:5b8a131321a845efa691d58309504a90/?field=title&field=event.type&field=project&field=user.display&field=timestamp&field=user.email&homepage=true&id=18214&name=&project=5428557&query=user.email%3Aphilipp%40example.com&sort=-timestamp&statsPeriod=1h&topEvents=5&yAxis=count%28%29).

**Comment by philipphofmann** [comment]
For profiling the app launch https://github.com/getsentry/sentry-cocoa/issues/3380 we need to start the profiler before SDK init. We likely need to store settings to disk and read them to find out if we should start the profiler. We could do something similar for app start spans. Depending on the setting we could swizzle methods of the AppDelegate, for example, but I'm unsure if it's worth the risk. WDTY, @armcknight and @brustolin?

**Comment by brustolin** [comment]
We could use a special entry in the app info.plist to setup this behaviour, it would start since the first launch. 
But I don't have a problem with using an extra setting file.

**Comment by philipphofmann** [comment]
I closed this issue for now. We get one extra app start span with https://github.com/getsentry/sentry-cocoa/pull/3534, which splits the UIKit and Application Init span into two spans. The Application Init span is a span users can impact significantly cause it shows how much time the app spends in didFinishWithLaunching. We also merged https://github.com/getsentry/sentry-cocoa/pull/3530, which ends the app start when the first frame is drawn. This will be default behavior in the next major version. I don't think swizzling early in the app start process is worth the risk at the moment with the other improvements we shipped so far. Furthermore, we can revisit this decision once we shipped app start profiling and reuse some of the logic to add extra spans.
